library( ape ) 
library( treedater ) 

.mltr <-  function( fn ){
	file.copy( fn, getwd() )
	d <- read.dna( fn, format='fasta' ) 
	#fn = basename( path_to_align )
	#tdir = tempdir() 
	#file.copy( fn, tdir ) 
	success = system( paste0( 'iqtree -redo -m HKY -s ', fn ), intern=FALSE  )
	if ( success != 0 ){
		stop('iqtree problem. Is it installed? ')
	}
	#file.copy( paste0( tdir, '/',  paste0( fn, '.treefile' ) ) , '.')
	mltr = read.tree( paste0( fn, '.treefile' ) )
	
	sids0 <- sapply( rownames(d) , function( x ){
	y = strsplit( x, split = '[\\/_\\|]+')[[1]]
	rey = paste( y, collapse = '([\\/_\\|]+)' ) 
		mltr$tip.label [ grepl( mltr$tip.label, pattern = rey ) ]
	})
	if ( (!is.vector( sids0 )) | (length( sids0 ) != Ntip(mltr ) )){
		print( setdiff( sids0, mltr$tip.label ))
		print( setdiff( mltr$tip.label, sids0 ))
		stop( 'Some tip labels in tree could not be matched sequence ids' )
	}
	
	sids <- setNames( rownames(d), sids0 )

	tr <- mltr
	tr$tip.label <- sids [ tr$tip.label ]
	
	tr 
}

#' Generate a starting tree for beast
#'
#' This assumes the sequence data are in the format generated by prep_tip_labels_phydyn
#' Requirements: 
#' 1) Fasta must have labels matching this form hCoV-19/Netherlands/Haarlem_1363688/2020|EPI_ISL_413572|2020-03-01|2020.16393442623|_Il
#' 2) iqtree must be installed and in the environment 
#' 
#' @param fn File name of input alignment. Must be in current working directory 
#' @param treeofn File name for output tree (newick)
#' @param ncpu Integer number of cores 
#' @return Will write the starting tree and returns the treedater tree 
#' @export 
make_starting_tree = function( fn , treeofn = 'startTree.nwk', ncpu = 4){
	tr <- .mltr( fn )

	sts <- sapply( strsplit( tr$tip.label, '\\|' ), function(x){
		as.numeric( tail(x,2)[1] )
	})
	names(sts) <- tr$tip.label 

	tr1 <- tr 
	tr1$edge.length <- pmax( 1/29e3/10 , tr1$edge.length )
	td = dater( unroot(tr1), sts, s= 29e3, omega0 = .0012, numStartConditions=0, meanRateLimits = c( .0009, .0015) , ncpu = ncpu )

	write.tree( td, file = treeofn ) 
	td 
}

#~ td = make_starting_tree( 'nl2.fasta' )

#' Make starting trees and insert into beast xml 
#' 
#' This will generate multiple starting trees by ML & treedater. 
#' Each tree is produced by a different random resolution of polytomies in the ML tree
#' Sequence names must have sample times included (see prep_tip_labels)
#' 
#' @param xmlfn File name of beast xml
#' @param fastafn File name of sequence data (needed for ML tree estimation)
#' @param ntres integer how many start trees to produce? a distinct xml is made for each 
#' @param ncpu Number of CPUs to use 
#' @return Some treedater trees. New XMLs are written to disk 
#' @export
add_starting_trees_to_xml <- function( xmlfn ,  fastafn , ntres = 1,  ncpu = 4 ){
	if ( inherits( fastafn, 'phylo' )  )
		tr <- fastafn 
	else
		tr = .mltr( fastafn  )
	
	sts <- sapply( strsplit( tr$tip.label, '\\|' ), function(x){
		as.numeric( tail(x,2)[1] )
	})
	names(sts) <- tr$tip.label 
	
	tr <- di2multi( tr, tol = 1e-5 ) # make polytomies 
	tres <- lapply( 1:ntres, function(i) { # resolve polytomies randomly 
		tr = unroot( multi2di( tr )  )  
		tr$edge.length <- pmax( 1e-6, tr$edge.length / 29e3 ) # note branch length is subst per genome ; we translate it to subst per site 
		tr
	})
	tds <- lapply( tres, function(tr){
		dater( unroot(tr), sts[tr$tip.label], s= 29e3, omega0 = .0012, numStartConditions=0, meanRateLimits = c( .0009, .0015) , ncpu = ncpu )
	})
	
	x = readLines( xmlfn ) 
	for ( k in 1:ntres ){
		xk = gsub( x , pattern = 'START_TREE', replacement = write.tree( tds[[k]] )  )
		if ( !grepl( pattern = '\\.xml$', xmlfn )  )
			writeLines( xk, con =  paste0( xmlfn, '.', k )  )
		else 
			writeLines( xk, con =  gsub( pattern = '\\.xml$', replacement = paste0('\\.',k,'\\.xml'), xmlfn ) )
	}
	tds
}
#~ xmlfn = 'nltest.xml' 
#~ tds = add_starting_trees_to_xml( xmlfn, 'nl2test.fasta', ntres = 2 )
